tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/TensorScope.java

```diff
@@ -0,0 +1,536 @@
+ limitations under the License.
+ ==============================================================================
+ */
+package org.tensorflow;
+
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+
+/**
+ * A scope that can be used to manage tensor resources.  Any tensors created between a scope's creation and calling
+ * {@code close()} that haven't been detached or attached to a different scope are guaranteed to be closed with the
+ * scope (even if they are created in a sub-scope).  Tensors may be manually closed earlier without issue.
+ * <p>
+ * Tensors are automatically tracked on creation.  A tensor can me manually added to a scope with {@link
+ * TensorScope#attach(Tensor)} or {@link Tensor#attachToCurrent()}.  A tensor may only have one scope: if it currently
+ * has a scope when {@code attach} is called, it is removed from its original scope.
+ * <p>
+ * {@link Tensor#detach()} detaches the tensor from it's scope, requiring the user to close it manually or attach it to
+ * another scope.
+ * <p>
+ * Scopes will be inherited at thread creation, but further scope creation on different threads will be independent,
+ * other than having the same parent. Closing a scope will close it's children regardless of which threads they are on.
+ */
+public class TensorScope implements AutoCloseable {
+
+  private static final InheritableThreadLocal<TensorScope> currentScope = new InheritableThreadLocal<>();
+
+  public static TensorScope currentScope() {
+    TensorScope scope = currentScope.get();
+
+    if (scope == null || !scope.closed) {
+      return scope;
+    }
+
+    // scope could be closed in another thread, in which case this thread's currentScope wouldn't be updated
+    while (scope != null && scope.closed) {
+      scope = scope.parent;
+    }
+    currentScope.set(scope);
+    return scope;
+  }
+
+  /**
+   * Runs {@code block}, then closes any tensors created during its execution.
+   * <p>To release tensors, use {@link #withCleanup(Consumer)} or one of the {@code produceWithCleanup} methods.
+   */
+  public static void withCleanup(Runnable block) {
+    try (TensorScope scope = new TensorScope()) {
+      block.run();
+    }
+  }
+
+  /**
+   * Runs {@code block}, then closes any tensors created during its execution (or attached to the scope).
+   * <p>Tensors can be released using the passed scope.
+   */
+  public static void withCleanup(Consumer<TensorScope> block) {
+    try (TensorScope scope = new TensorScope()) {
+      block.accept(scope);
+    }
+  }
+
+  /**
+   * Runs {@code block} and returns the result, then closes any tensors created during its execution.
+   * <p>To release tensors, use {@link #withCleanup(Function)} or one of the {@code produceWithCleanup} methods.
+   */
+  public static <T> T withCleanup(Supplier<T> block) {
```

Comment:
I'm tempted to check if the result here is `Tensor`, `HasTensor`, or `Iterable<Tensor>` and release it if so, rather than having distinct methods.  The only thing stopping me is that the behavior could change based on the result if you're returning a supertype of any of those types.  It doesn't seem worth it to me, thoughts?